zpwr/man/zpwr.1
.\" Copyright 2025 IBM Corp.
.\" s390-tools is free software; you can redistribute it and/or modify
.\" it under the terms of the MIT license. See LICENSE for details.
.\"
.ds c \fBzpwr\fP
.TH zpwr 1 "Feb 2025" "s390-tools" zpwr
.SH NAME
zpwr \- Display power readings of a partition and central processor complex
(CPC) on an IBM Z or LinuxONE server.
.SH SYNOPSIS
.sp
.nf
\fIzpwr\fR
	[\-d <NUMBER> | \-\-delay=<NUMBER>] [\-c <NUMBER> | \-\-count=<NUMBER>]
	[\-\-format=<STRING>]
	[\-s | \-\-stream]
	[\-v | \-\-version]
	[\-h | \-\-help]
.fi
.SH DESCRIPTION
.sp
zpwr displays power readings of a partition and central processor complex
(CPC) on an IBM Z or LinuxONE server.
.sp
Partition readings include CPU, storage and I/O power readings.  It specifies
the number of units allocated to the configuration at the time the reading was
taken.
.sp
CPC readings include total, unassigned resources and infrastructure power
readings.
.sp
Power reading units are automatically presented in a human readable format,
whereas complex formats like json, csv, json-seq, jsonl, pairs display the
power readings in microwatts.
.sp
CPC power readings can be retrieved only when the required authorization was
set for the LPAR.  Enable the required authorization by setting "Global
Performance Data" in LPAR's activation profile.
.sp
If the power readings cannot be retrieved, the power readings return an empty
string or null in complex formats, and the power readings are not displayed in
human readable format.
.SH OPTIONS
.PP
\-\-format=<STRING>
.RS 4
Retrieve power readings in the following formats: json, csv,
json-seq or pairs. If no format is specified in the command, the output will
default to the human readable format.
.sp
When the format is json, csv, json-seq, jsonl or pairs, the power readings are
displayed in microwatts.
.RE
.PP
\-d, \-\-delay
.RS 4
Retrieve power readings after specified delay in seconds has elapsed.
.RE
.PP
\-c <NUMBER>, \-\-count=<NUMBER>
.RS 4
Required number of power readings.
.sp
The \-\-count option can only be used in conjunction with \-\-stream or
\-\-delay option.
.RE
.PP
\-s, \-\-stream
.RS 4
Retrieve power readings in stream mode. A new power reading will be displayed
only after an interval has elapsed since the last power reading request.
The interval is automatically calculated in stream mode.
.sp
The \-\-stream and \-\-delay options are mutually exclusive.
.RE
.PP
\-v, \-\-version
.RS 4
Print version information.
.RE
.PP
\-h, \-\-help
.RS 4
Print usage information.
.RE
.SH OUTPUT
.B Iteration
.RS 4
Iteration number of the current power readings. If the power reading is
requested before the time interval has elapsed, cached power readings might be
returned. The iteration number helps to determine if the cached power readings
are provided.
.RE
.PP
.B Time
.RS 4
The time at which power readings were taken, expressed in human readable format.
.RE
.PP
.B Time epoch
.RS 4
The time at which power readings were taken, expressed in epoch time.
.PP
.B sec
.RS 4
The number of seconds elapsed since the Unix epoch.
.RE
.PP
.B nsec
.RS 4
The number of nanoseconds elapsed since the last full seconds.
.RE
.RE
.PP
.B Update interval
.RS 4
Indicates the time interval in nanoseconds after which new power readings are
available. If the power readings is requested before the time interval has
elapsed, cached power readings might be returned.
.RE
.PP
.B CPU
.RS 4
CPU power reading of a partition.
.RE
.PP
.B Storage
.RS 4
Storage power reading of a partition.
.RE
.PP
.B I/O
.RS 4
I/O power reading of a partition.
.RE
.PP
.B Total
.RS 4
Specifies number of units for all electrical and mechanical components of the
system.
.RE
.PP
.B Unassigned Resources
.RS 4
Specifies number of units for all types of resources in the standby state or
reserved state at the time the reading was taken.
.RE
.PP
.B Infrastructure
.RS 4
Specifies number of units for all subsystems in the CPC which do not provide
CPU, storage or I/O resources to partitions.
.RE
.SH "EXAMPLES"
1. Display power readings in human readable format.
.nf
.ft CR
# zpwr
LPAR CPU:                       140.00  W
LPAR Storage:                     1.72  W
LPAR I/O:                        40.00  W

CPC Total:                       15.40 kW
CPC Unassigned Resources:         7.48 kW
CPC Infrastructure:               5.31 kW

Update interval: 10.00 s
.ft
.fi
.sp 1
2. Perform two power measurements with a delay of 10 seconds and output the
data in json format.
.nf
.ft CR
# zpwr --format json --delay 10 --count 2
{
  "meta": {
    "api_level": "1",
    "version": "2.36.0-build-20250107",
    "host": "lpar001",
    "time_epoch": "1736316898",
    "time": "2025-01-08 07:14:58+0100"
  },
  "zpwr": [
    {
      "iteration": "0",
      "time": "2025-01-08 07:14:58+0100",
      "time_epoch_sec": "1736316898",
      "time_epoch_nsec": "10935887",
      "update_interval": "10000000000",
      "lpar": [
        "cpu": "140000000",
        "storage": "1719200",
        "io": "40000000"
      ],
      "cpc": [
        "total": "15406000000",
        "unassigned_resources": "7578848100",
        "infrastructure": "5283946300"
      ]
    },
    {
      "iteration": "1",
      "time": "2025-01-08 07:15:08+0100",
      "time_epoch_sec": "1736316908",
      "time_epoch_nsec": "11113153",
      "update_interval": "10000000000",
      "lpar": [
        "cpu": "140000000",
        "storage": "1719200",
        "io": "40000000"
      ],
      "cpc": [
        "total": "15401000000",
        "unassigned_resources": "7569928400",
        "infrastructure": "5284866000"
      ]
    }
  ]
}
.ft
.fi
.sp 1
3. Display power readings in csv format and in stream mode.
.nf
.ft CR
# zpwr --format csv --stream
"iteration","time","time_epoch_sec","time_epoch_nsec","update_interval","cpu","storage","io","total","unassigned_resources","infrastructure"
"0","2025-01-08 07:17:05+0100","1736317025","592784684","10000000000","143000000","1708200","40000000","15550000000","7659162700","5337186700"
"1","2025-01-08 07:17:15+0100","1736317035","592937261","10000000000","143000000","1708200","40000000","15533000000","7650389100","5328960300"
.ft
.fi


CHANGELOG.md
Release history for s390-tools (MIT version)
--------------------------------------------

* __v2.xx.x (20xx-xx-xx)__

  For Linux kernel version: 6.x

  Add new tools / libraries:

  Changes of existing tools:

  Bug Fixes:

* __v2.41.0 (2026-02-16)__

  For Linux kernel version: 6.19

  Changes of existing tools:
  - chreipl: Make --bootparms work for ECKD re-IPL
  - cpacfstats: Add 'unauthorized' state to CPU-MF counters
  - cpictl: Detect RHCOS using VARIANT_ID
  - hsci: Automatically set appropriate MTU for HSCI
  - libutil: Add util_readlink() and util_readlinkat() helpers
  - libutil: Add util_startswith() to util_str
  - libutil: Add utility parsing functions
  - lschp: Add support for structured output (--format)
  - lsreipl: Suppress 'clear' output if not supported
  - pvimg: Add '--format text' support to 'pvimg info'
  - pvimg: Add '--print-schema <FORMAT>' option to 'pvimg info'
  - pvimg: Add '--show-secrets' flag to 'pvimg info'
  - pvimg: Provide improved JSON output to 'pvimg info --format json'
  - pvinfo: Improve User experience on non-SE enabled systems
  - zipl/ngdump: Ensure ext4 file system is used on dump partition
  - zkey: Add support for integrity protected disks using HMAC keys

  Bug Fixes:
  - cpumf/pai: Handle different size of perf_event_attr
  - lscss: Fix memory leak
  - zipl: Fix dump job on tape devices

* __v2.40.0 (2025-12-11)__

  For Linux kernel version: 6.18

  Add new tools / libraries:
  - Add project-wide .clang-tidy configuration
  - libutil: Introduce util_time for time related functionality
  - libutil: Introduce zsh/bash autocompletion tooling based on util_opt
  - pvinfo: Tool to display Secure Execution system information
  - pvverify: Tool to verify host-key documents

  Changes of existing tools:
  - cpumf: Implement zsh and bash autocompletion
  - dasdfmt: Implement zsh and bash autocompletion
  - dbginfo.sh: Add NetworkManager and netplan
  - dbginfo.sh: Add kvm_stat
  - dbginfo.sh: Adding stp time information
  - dbginfo.sh: Simplify procfs collection
  - hyptop: Add physical information row
  - hyptop: Calculate sample time delta for physical partition
  - hyptop: Replace long option names using _ with - for consistency
            For example: --cpu_types -> --cpu-types
            (Options with _ are still supported for backward compatibility)
  - libekmfweb: Add function to validate a certificate against the identity key
  - netboot: Add longer kernel command lines support
  - udev/rules.d: Make virtio-blk devices non-rotational
  - udev/rules.d: Set default io scheduler to 'none' for virtio-blk
  - ziomon: Add support to sample device symlinks (/dev/disk/...)
  - ziorep_config: Add fcp-lun details to -M option output
  - ziorep_config: Add port_id and failed attributes to -A option output
  - netboot: Install on non-s390 architectures

  Bug Fixes:
  - lib(ekmfweb|kmipclient): Use ln without -r
  - s390-tools: Fix various compilation issues with musl libc
  - zipl/boot: Fix unused loadparm when SCLP line-mode console is absent

* __v2.39.0 (2025-10-14)__

  For Linux kernel version: 6.16 / 6.17

  Changes of existing tools:
  - chpstat: Add options to select IEC units for scaling (SI units are default)
  - chzdev: Introduce --no-module-load option
  - cpi: Disable CPI for SEL guests by default
  - dbginfo.sh: Enhance logging on timeout triggered
  - iucvterm: Install symlink for lsiucvallow.8 man page
  - lshwc: Add command line flag to specify individual counters
  - lspai: Add command line flag for delta values
  - lspai: Add command line flag for short counter names
  - lspai: Add command line flag to specify individual counters
  - lspai: Add command line flags for all cpus
  - lspai: Add command line flags for hexadecimal output
  - man: Use CR for constant width font
  - pvimg: Add '--image-key' option
  - zdev: Allow dynamic control of module load
  - zipl/boot: Fix EBCDIC code page 500 conversion and decrease size by 200 bytes
  - zipl: Add support of heterogeneous mirrors (remove technical limitations
    on mirrored targets, thus allowing mirrored devices consist of partitions
    at different offsets on disks of different types and geometry).
  - zkey: Add support for generating and importing exportable secure keys

  Bug Fixes:
  - chpstat: Fix scaling of DPU utilization calculation
  - zdev/dracut: Prevent loading of unused kernel modules
  - zdev: Fix double device configuration on DPM systems
  - zdev: Fix double device configuration with rd.dasd
  - zipl_helper.device-mapper: Fix segfault in an error path

* __v2.38.0 (2025-06-25)__

  For Linux kernel version: 6.14 / 6.15

  Add new tools:
  - udev: New rule to set newly hotplugged CPUs online
  - zmemtopo: Display memory topology information
  - zpwr: Display power readings of a partition and CPC

  Removed tools / features:
  - check_hostkeydoc: Remove installation target
  - scsi_logging_level: Delete SCSI logging script (available in sg3_utils)
  - zdump: Drop build_arch for s390 DASD dumps
  - zdump: Drop non-extended multi-volume DASD dump support
  - zdump: Drop support of 32-bit dump architecture
  - zdump: Drop support of non-extended single volume DASD dumpers
  - zdump: Drop support of obsolete dumps and dumpers

  Changes of existing tools / libraries:
  - Various man-pages fixes
  - check_hostkeydoc: Add deprecation warning
  - check_hostkeydoc: Move to scripts directory
  - cpuplugd: Allow cpu hotplugging on systems without polarization
  - dbginfo.sh: Add Ubuntu snap tool
  - dbginfo.sh: Add missing config data and logs
  - dbginfo.sh: Reworking the container section
  - dbginfo.sh: Update for network commands
  - dbginfo.sh: Updating info for disks and lvm
  - libutil: Add machine type definition for machines 9175 and 9176
  - lscpumf: Add support for IBM z17 counter sets
  - lshwc: Add command line flag for run time
  - lshwc: Add flags to display counter values in hex
  - lshwc: Add output '--format' option

zpwr/zpwr.c
/*
 * zpwr - Display power readings of s390 computing environment.
 *
 * Display power readings for resources in s390 computing environment from
 * power information block (pib).
 *
 * Copyright IBM Corp. 2025
 *
 * s390-tools is free software; you can redistribute it and/or modify
 * it under the terms of the MIT license. See LICENSE for details.
 */

#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <iconv.h>
#include <math.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <time.h>

#include "lib/util_base.h"
#include "lib/util_fmt.h"
#include "lib/util_opt.h"
#include "lib/util_prg.h"
#include "zpwr.h"

#define DIAG		"/dev/diag"
#define NANO		1000000000ULL
#define NUMUNIT		5
#define NAMELEN		8
#define COMPWIDTH	27
#define OPT_FORMAT	256

enum part_power {
	CPU,
	STORAGE,
	IO,
	MAX_PM_PARTITION,
};

enum cpc_power {
	TOTAL,
	UNASSIGNED,
	INFRA,
	MAX_PM_CPC,
};

struct zpwrinfo {
	u64 part[MAX_PM_PARTITION];
	u64 cpc[MAX_PM_CPC];
	bool pvalid;
	bool cvalid;
};

static const char *simplefmt_part[MAX_PM_PARTITION] = {
	"LPAR CPU:",
	"LPAR Storage:",
	"LPAR I/O:",
};

static const char *simplefmt_cpc[MAX_PM_CPC] = {
	"CPC Total:",
	"CPC Unassigned Resources:",
	"CPC Infrastructure:",
};

static const char *complexfmt_part[MAX_PM_PARTITION] = {
	"cpu",
	"storage",
	"io",
};

static const char *complexfmt_cpc[MAX_PM_CPC] = {
	"total",
	"unassigned_resources",
	"infrastructure",
};

static const struct util_prg prg = {
	.desc = "Power readings of s390 computing environment",
	.copyright_vec = {
		{
			.owner = "IBM Corp.",
			.pub_first = 2025,
			.pub_last = 2025,
		},
		UTIL_PRG_COPYRIGHT_END
	}
};

static struct util_opt opt_vec[] = {
	UTIL_OPT_SECTION("OPTIONS"),
	{
		.option = { "format", required_argument, NULL, OPT_FORMAT },
		.argument = "FORMAT",
		.flags = UTIL_OPT_FLAG_NOSHORT,
		.desc = "List data in specified FORMAT (" FMT_TYPE_NAMES ")",
	},
	{
		.option = { "delay", required_argument, NULL, 'd' },
		.argument = "NUMBER",
		.desc = "Power readings after delay (seconds)",
	},
	{
		.option = { "count", required_argument, NULL, 'c' },
		.argument = "NUMBER",
		.desc = "Number of power readings",
	},
	{
		.option = { "stream", no_argument, NULL, 's' },
		.desc = "Power readings in stream mode",
	},
	UTIL_OPT_HELP,
	UTIL_OPT_VERSION,
	UTIL_OPT_END
};

static int get_max_column_width(struct zpwrinfo *pinfo)
{
	u64 max = 0;
	int i, col;

	for (i = 0; i < MAX_PM_PARTITION; i++) {
		if (pinfo->part[i] > max)
			max = pinfo->part[i];
	}
	for (i = 0; i < MAX_PM_CPC; i++) {
		if (pinfo->cpc[i] > max)
			max = pinfo->cpc[i];
	}
	col = (int)log10((double)max) + 1;
	/* power unit and space consideration */
	col += 3;
	return col;
}

static char *get_human_readable_unit(u64 val)
{
	const char *unitstr[NUMUNIT] = { "uW", "mW", " W", "kW", "MW" };
	int exponent[NUMUNIT] = { 1, 3, 6, 9, 12 }, unitindex = 0, i;
	double res, smallestres = (double)val;
	char *buf;

	for (i = 1; i < NUMUNIT; i++) {
		res = (double)val / pow(10, exponent[i]);
		if ((u64)res && res < smallestres) {
			smallestres = res;
			unitindex = i;
		}
	}
	util_asprintf(&buf, "%.2f %s", smallestres, unitstr[unitindex]);
	return buf;
}

static double get_human_readable_interval(u64 val, bool *seconds)
{
	double res;

	res = (double)val / pow(10, 9);
	if ((u64)res)
		*seconds = true;
	else
		*seconds = false;
	return *seconds ? res : (double)val;
}

/* From linux arch/s390/include/asm/timex.h */
static unsigned long tod_to_ns(unsigned long todval)
{
	return ((todval >> 9) * 125) + (((todval & 0x1ff) * 125) >> 9);
}

static void reset_zpwrinfo(struct zpwrinfo *pinfo, struct pib *pib,
			   unsigned long buffersize)
{
	memset(pinfo, 0, sizeof(*pinfo));
	memset(pib, 0, buffersize);
}

static void print_zpwrinfo(struct zpwrinfo *pinfo, u64 iteration,
			   int fmt_specified, u64 interval)
{
	enum util_fmt_mflags_t fmt_mflags = FMT_DEFAULT;
	bool secondflag = false;
	struct timespec ts;
	char timestr[30];
	char *simplestr;
	int i, colwidth;
	struct tm *tm;

	if (!fmt_specified) {
		colwidth = get_max_column_width(pinfo);
		for (i = 0; i < MAX_PM_PARTITION; i++) {
			if (!pinfo->pvalid)
				break;
			printf("%-*s", COMPWIDTH, simplefmt_part[i]);
			simplestr = get_human_readable_unit(pinfo->part[i]);
			printf("%*s\n", colwidth, simplestr);
			free(simplestr);
		}
		printf("\n");
		for (i = 0; i < MAX_PM_CPC; i++) {
			if (!pinfo->cvalid)
				break;
			printf("%-*s", COMPWIDTH, simplefmt_cpc[i]);
			simplestr = get_human_readable_unit(pinfo->cpc[i]);
			printf("%*s\n", colwidth, simplestr);
			free(simplestr);
		}
		printf("\n");
		printf("Update interval: %.2f %s\n",
		       get_human_readable_interval(interval, &secondflag),
		       secondflag ? "s" : "ns");
		return;
	}
	clock_gettime(CLOCK_REALTIME, &ts);
	tm = localtime(&ts.tv_sec);
	strftime(timestr, sizeof(timestr), "%F %T%z", tm);
	util_fmt_obj_start(FMT_ROW, "iteration");
	util_fmt_pair(fmt_mflags, "iteration", "%llu", iteration);
	util_fmt_pair(fmt_mflags, "time", "%s", timestr);
	util_fmt_pair(fmt_mflags, "time_epoch_sec", "%lld", ts.tv_sec);
	util_fmt_pair(fmt_mflags, "time_epoch_nsec", "%ld", ts.tv_nsec);
	util_fmt_pair(fmt_mflags, "update_interval", "%llu", interval);
	util_fmt_obj_start(FMT_LIST, "lpar");
	for (i = 0; i < MAX_PM_PARTITION; i++)
		util_fmt_pair(pinfo->pvalid ? fmt_mflags : fmt_mflags | FMT_INVAL,
			      complexfmt_part[i], "%llu", pinfo->part[i]);
	util_fmt_obj_end(); /* End of lpar list */
	util_fmt_obj_start(FMT_LIST, "cpc");
	for (i = 0; i < MAX_PM_CPC; i++)
		util_fmt_pair(pinfo->cvalid ? fmt_mflags : fmt_mflags | FMT_INVAL,
			      complexfmt_cpc[i], "%llu", pinfo->cpc[i]);
	util_fmt_obj_end(); /* End of cpc list */
	util_fmt_obj_end(); /* End of iteration row */
}

static int read_zpwrinfo(struct zpwrinfo *pinfo, struct pib *pib)
{
	struct pib_prologue *prologue;
	int i, comp, max = 0, rc = 0;
	u64 *curr_zpwrinfo;
	u8 *metrics;
	void *ptr;

	ptr = (u8 *)pib + pib->hlen;
	prologue = ptr;
	for (i = 0; i < pib->num; i++) {
		metrics = (u8 *)prologue + sizeof(*prologue);
		if (prologue->format == 0) {
			curr_zpwrinfo = pinfo->part;
			max = MAX_PM_PARTITION;
			pinfo->pvalid = true;
		} else if (prologue->format == 1) {
			curr_zpwrinfo = pinfo->cpc;
			max = MAX_PM_CPC;
			pinfo->cvalid = true;
		} else {
			rc = -EINVAL;
			warnx("Unknown format detected:%d\n", prologue->format);
			break;
		}
		metrics += NAMELEN;
		for (comp = 0; comp < max; comp++) {
			memcpy(&curr_zpwrinfo[comp], metrics, sizeof(u64));
			metrics += sizeof(u64);
		}
		ptr = (u8 *)prologue + prologue->len;
		prologue = ptr;
	}
	return rc;
}

static void fmt_start(enum util_fmt_t fmt, unsigned int fmt_flags,
		      int fmt_specified)
{
	if (!fmt_specified)
		return;
	util_fmt_init(stdout, fmt, fmt_flags, 1);
	if (!util_fmt_is_json_stream(fmt))
		util_fmt_obj_start(FMT_LIST, "zpwr");
}

static void fmt_end(enum util_fmt_t fmt, int fmt_specified)
{
	if (!fmt_specified)
		return;
	if (!util_fmt_is_json_stream(fmt))
		util_fmt_obj_end(); /* zpwr[] */
	util_fmt_exit();
}

int main(int argc, char *argv[])
{
	enum util_fmt_flags_t fmt_flags = FMT_HANDLEINT | FMT_QUOTEALL | FMT_KEEPINVAL;
	int ch, fd, rc = EXIT_FAILURE, fmt_specified = 0;
	bool stream = false, init = true;
	enum util_fmt_t fmt = FMT_JSON;
	u64 init_seq, interval = 0;
	long count = 0, delay = 0;
	struct diag324_pib data;
	struct zpwrinfo *pinfo;
	struct timespec ts;
	size_t buffersize;
	struct pib *pib;
	struct stat st;

	util_prg_init(&prg);
	util_opt_init(opt_vec, NULL);
	while (1) {
		ch = util_opt_getopt_long(argc, argv);
		if (ch == -1)
			break;
		switch (ch) {
		case 'c':
			errno = 0;
			count = strtoul(optarg, NULL, 0);
			if (errno || count <= 0)
				errx(EXIT_FAILURE, "Positive number expected for option -%c", ch);
			break;
		case 'd':
			errno = 0;
			delay = strtoul(optarg, NULL, 0);
			if (errno || delay <= 0)
				errx(EXIT_FAILURE, "Positive number expected for option -%c", ch);
			break;
		case 's':
			stream = true;
			break;
		case OPT_FORMAT:
			if (!util_fmt_name_to_type(optarg, &fmt)) {
				errx(EXIT_FAILURE, "Supported formats: %s", FMT_TYPE_NAMES);
			} else {
				if (fmt == FMT_CSV)
					fmt_flags |= FMT_NOMETA;
				else
					fmt_flags |= FMT_DEFAULT;
				fmt_specified = 1;
			}
			break;
		case 'h':
			util_prg_print_help();
			util_opt_print_help();
			return EXIT_SUCCESS;
		case 'v':
			util_prg_print_version();
			return EXIT_SUCCESS;
		default:
			util_opt_print_parse_error(ch, argv);
			return EXIT_FAILURE;
		}
	}
	if (stream && delay)
		errx(EXIT_FAILURE, "-s and -d option are mutually exclusive");
	if (count && !delay && !stream)
		errx(EXIT_FAILURE, "-c option can only be used in conjunction with -d or -s");
	if (stat(DIAG, &st) == -1)
		errx(EXIT_FAILURE, "Missing kernel support to retrieve power readings");
	fd = open(DIAG, O_RDONLY);
	if (fd < 0)
		err(EXIT_FAILURE, "Open failed: %s", DIAG);
	rc = ioctl(fd, DIAG324_GET_PIBLEN, &buffersize);
	if (rc && errno == EOPNOTSUPP) {
		warnx("The machine does not support retrieving power readings");
		goto out;
	} else if (rc) {
		warn("Ioctl (DIAG324_GET_PIBLEN) failed");
		goto out;
	}
	pinfo = calloc(1, sizeof(*pinfo));
	if (!pinfo) {
		warnx("Allocation of pinfo failed");
		goto out;
	}
	pib = calloc(1, buffersize);
	if (!pib) {
		free(pinfo);
		warnx("Allocation of pib failed");
		goto out;
	}
	data.address = (u64)pib;
	fmt_start(fmt, fmt_flags, fmt_specified);
	while (true) {
		rc = ioctl(fd, DIAG324_GET_PIBBUF, &data);
		if (rc != 0 && errno != EBUSY) {
			warn("Ioctl (DIAG324_GET_PIBBUF) failed");
			goto out_free;
		}
		rc = read_zpwrinfo(pinfo, pib);
		if (rc)
			goto out_free;
		if (init) {
			init_seq = data.sequence;
			init = false;
		}
		interval = tod_to_ns(pib->intv);
		print_zpwrinfo(pinfo, data.sequence - init_seq, fmt_specified, interval);
		reset_zpwrinfo(pinfo, pib, buffersize);
		if (stream) {
			ts.tv_sec = interval / NANO;
			ts.tv_nsec = interval % NANO;
		} else {
			ts.tv_sec = delay;
			ts.tv_nsec = 0;
		}
		if ((stream || delay) && !count) {
			nanosleep(&ts, NULL);
			continue;
		} else if (--count > 0) {
			nanosleep(&ts, NULL);
			continue;
		} else {
			break;
		}
	}
	fmt_end(fmt, fmt_specified);
out_free:
	free(pinfo);
	free(pib);
out:
	close(fd);
	return rc ? EXIT_FAILURE : EXIT_SUCCESS;
}
